topology/nhlt/nhlt-processor.c

```diff
@@ -0,0 +1,321 @@
+  it under the terms of version 2 of the GNU General Public License as
+  published by the Free Software Foundation.
+
+  This program is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
+  The full GNU General Public License is included in this distribution
+  in the file called LICENSE.GPL.
+*/
+
+#include <errno.h>
+#include <stdio.h>
+#include <alsa/input.h>
+#include <alsa/output.h>
+#include <alsa/conf.h>
+#include <alsa/error.h>
+#include "pre-processor.h"
+#include "nhlt-processor.h"
+#include "nhlt.h"
+
+#define MAX_ENDPOINT_COUNT 20
+
+static int add_bytes_to_data_section(struct tplg_pre_processor *tplg_pp, char *src)
+{
+	const char *data_name = "nhlt";
+	snd_config_t *top_data_section;
+	snd_config_t *bytes;
+	snd_config_t *nhlt;
+	int ret;
+
+	/* add the nhlt bytes as new config into existing SectionData*/
+	ret = snd_config_search(tplg_pp->output_cfg, "SectionData", &top_data_section);
+	if (ret < 0)
+		return ret;
+
+	ret = tplg_config_make_add(&nhlt, data_name, SND_CONFIG_TYPE_COMPOUND, top_data_section);
+	if (ret < 0)
+		return ret;
+
+	ret = tplg_config_make_add(&bytes, "bytes", SND_CONFIG_TYPE_STRING, nhlt);
+	if (ret < 0) {
+		snd_config_delete(nhlt);
+		return ret;
+	}
+
+	ret = snd_config_set_string(bytes, src);
+	if (ret < 0) {
+		snd_config_delete(nhlt);
+		snd_config_delete(bytes);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int add_to_manifest(struct tplg_pre_processor *tplg_pp, char *src)
+{
+	const char *data_name = "nhlt";
+	snd_config_t *manifest;
+	snd_config_t *man_inst;
+	snd_config_t *n;
+	snd_config_t *cfg_new;
+	snd_config_t *data;
+	snd_config_iterator_t i, next;
+	const char *id;
+	char *id_name;
+	int ret, id_num = 0;
+
+	/* find manifest section and its named instance */
+	ret = snd_config_search(tplg_pp->output_cfg, "SectionManifest", &manifest);
+	if (ret < 0)
+		return 0;
+
+	snd_config_for_each(i, next, manifest) {
+
+		n = snd_config_iterator_entry(i);
+
+		if (snd_config_get_id(n, &id) < 0)
+			continue;
+
+		if (!strcmp(id, "sof_manifest"))
+			man_inst = n;
+	}
+
+	/* if manifest not found, just exit */
+	if (!man_inst)
+		return 0;
+
+	ret = add_bytes_to_data_section(tplg_pp, src);
+	if (ret < 0)
+		return ret;
+
+	/* add the reference to the bytes into manifest data */
+	ret = snd_config_search(man_inst, "data", &data);
+	if (ret < 0)
+		return ret;
+
+	snd_config_for_each(i, next, data) {
+		const char *name;
+
+		n = snd_config_iterator_entry(i);
+		if (snd_config_get_string(n, &name) < 0)
+			continue;
+
+		/* item already exists */
+		if (!strcmp(name, data_name))
```

Comment:
how does this work? so you have one NHLT blob for types of DAI?

Reply:
it is just 1 acpi nhlt blob, that might contain 1 to many nhlt endpoints (DAI configs). So I'm just trying to add 1 data element into manifest, nothing more. I think the code is a copy from some other part from topology2 processing, can be changed if there's standard way of adding data element.